pkg/datapath/linux/node.go

```diff
@@ -947,9 +952,278 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+				continue
+			}
+
+			mtu := n.nodeConfig.MtuConfig.GetDeviceMTU()
+			if err := netlink.LinkSetMTU(link, mtu); err != nil {
+				log.WithError(err).Errorf("ip link set %d mtu %d failed", addedInterface.Index, mtu)
+			}
+
+			if err := netlink.LinkSetUp(link); err != nil {
+				log.WithError(err).Errorf("ip link set %d up failed", addedInterface.Index)
+			}
+		}
+
+		// Ignore removed interfaces for now.
+		_ = removedInterfaces
+
+		oldRules, oldRoutes := nodeRulesAndRoutes(oldNode)
+		newRules, newRoutes := nodeRulesAndRoutes(newNode)
+		addedRules, removedRules := diffRules(oldRules, newRules)
+		addedRoutes, removedRoutes := diffRoutes(oldRoutes, newRoutes)
+
+		// Add and remove rules and routes. This has to succeed so we retry
+		// multiple times.
+		maxRetries := 3
+		rulesToAdd, rulesToRemove := addedRules, removedRules
+		routesToAdd, routesToRemove := addedRoutes, removedRoutes
+		var failedAddRules, failedRemoveRules []*route.Rule
+		var failedAddRoutes, failedRemoveRoutes []*netlink.Route
+		for retry := 0; retry < maxRetries; retry++ {
+			for _, rule := range rulesToAdd {
+				if err := route.ReplaceRule(*rule); err != nil {
+					log.WithError(err).Errorf("add rule %s failed", rule)
+					failedAddRules = append(failedAddRules, rule)
+				}
+			}
+
+			for _, rule := range rulesToRemove {
+				if err := route.DeleteRule(*rule); err != nil {
+					log.WithError(err).Errorf("delete rule %s failed", failedRemoveRules)
+					failedRemoveRules = append(failedRemoveRules, rule)
+				}
+			}
+
+			for _, route := range routesToAdd {
+				if err := netlink.RouteReplace(route); err != nil {
+					log.WithError(err).Errorf("add L2 nexthop route %s failed", err)
+					failedAddRoutes = append(failedAddRoutes, route)
+				}
+			}
+
+			for _, route := range routesToRemove {
+				if err := netlink.RouteDel(route); err != nil {
+					log.WithError(err).Errorf("remove L2 nexthop route %s failed", err)
+					failedRemoveRoutes = append(failedAddRoutes, route)
+				}
+			}
+
+			// If there were no failues, then we are done.
+			if len(failedAddRules)+len(failedRemoveRules)+len(failedAddRoutes)+len(failedRemoveRoutes) == 0 {
+				break
+			}
+
+			// Otherwise, retry with the failures and clear the list of failures.
+			rulesToAdd, failedAddRules = failedAddRules, nil
+			rulesToRemove, failedRemoveRules = failedRemoveRules, nil
+			routesToAdd, failedAddRoutes = failedAddRoutes, nil
+			routesToRemove, failedRemoveRoutes = failedRemoveRoutes, nil
+		}
+
+		// If there were still failures after retrying, then return an error.
+		if failures := len(failedAddRules) + len(failedRemoveRules) + len(failedAddRoutes) + len(failedRemoveRoutes); failures > 0 {
+			return fmt.Errorf("adding and removing %d rules and routes failed after %d retries", failures, maxRetries)
+		}
+	}
+
 	return nil
 }
 
+// nodeRulesAndRoutes returns the rules and routes required to configure node.
+// It is based on pkg/datapath/linux/routing.Configure.
```

Comment:
When you say "based on", what is the overlap and how can we minimize code duplication?

Reply:
`Configure` is still used for `ipam: azure`. The amount of overlap depends on the configuration. I would suggest that consolidating the code is best done in a follow-up PR.